packages/eventual-send/src/index.js

```diff
@@ -71,33 +68,109 @@ export function makeHandledPromise() {
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.delete(target);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         promiseToPresence.set(target, presence);
         target = parent;
       }
     } else {
-      // We propagate p and remove all other unsettled handlers
+      // We propagate p and remove all other pending handlers
       // upstream.
       // Note that everything except presences is covered here.
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.set(target, p);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         target = parent;
       }
     }
     return target;
   }
 
-  // This special handler accepts Promises, and forwards
-  // handled Promises to their corresponding fulfilledHandler.
+  /**
+   * This special handler accepts Promises, and forwards
+   * handled Promises to their corresponding fulfilledHandler.
+   *
+   * @type {Required<EHandler<any>>}
+   */
   let forwardingHandler;
   let handle;
 
-  function HandledPromise(executor, unsettledHandler = undefined) {
-    if (new.target === undefined) {
-      throw new Error('must be invoked with "new"');
+  /**
+   * @param {string} handlerName
+   * @param {EHandler<any>} handler
+   * @param {string} operation
+   * @param {any} o
+   * @param {any[]} opArgs
+   * @param {Promise<unknown>} [returnedP]
+   * @returns {any}
+   */
+  const dispatchToHandler = (
+    handlerName,
+    handler,
+    operation,
+    o,
+    opArgs,
+    returnedP,
+  ) => {
+    let actualOp = operation;
+    const triedOperations = [operation];
+    const isSendOnly = actualOp.endsWith('SendOnly');
+    const makeResult = result => (isSendOnly ? undefined : result);
+
+    if (isSendOnly) {
+      // We don't specify the resulting promise if it is sendonly.
+      returnedP = undefined;
+    }
+
+    if (isSendOnly && typeof handler[actualOp] !== 'function') {
+      // Substitute for sendonly with the corresponding non-sendonly operation.
+      actualOp = actualOp.slice(0, -'SendOnly'.length);
+      triedOperations.push(actualOp);
+    }
+
+    // Fast path: just call the actual operation.
+    if (typeof handler[actualOp] === 'function') {
```

Comment:
Yes, you're right to remind me.  I intentionally violated that principle only because I forgot about `Function.prototype.apply` and called the method as `handler[actualOp](...args)`, which you'll now probably remind me how to uncurry properly.

Reply:
> which you'll now probably remind me how to uncurry properly.

Aha!  `Reflect.apply` does the trick.